### PR TITLE
[Issue 362] Improve Performance constructing the `asofJoin()`

### PR DESCRIPTION
**Pull request by tnixon**
Fix for #362 - improving performance of constructing the `asofJoin(...)` function.

### CODE REVIEW

**Review by rportilla-databricks — APPROVED** [review]
LGTM

File: python/tempo/tsdf.py [hunk 1]
```diff
@@ -157,38 +155,39 @@ def __addPrefixToColumns(self, col_list: list[str], prefix: str) -> "TSDF":
...
 
-        df = reduce(
-            lambda df, idx: df.withColumnRenamed(
-                col_list[idx], "".join([prefix, col_list[idx]])
-            ),
-            range(len(col_list)),
-            self.df,
+        # build a column rename map
+        col_map = {col: "_".join([prefix, col]) for col in col_list}
+        # df = self.df.withColumnsRenamed(col_map)
```
Comment (R7L208) [inline python/tempo/tsdf.py:164]:
```suggestion
```

File: python/tempo/tsdf.py [hunk 2]
```diff
@@ -157,38 +155,39 @@ def __addPrefixToColumns(self, col_list: list[str], prefix: str) -> "TSDF":
...
+
+        # build a list of column expressions to rename columns in a select
```
Comment (R7L208) [inline python/tempo/tsdf.py:166]:
```suggestion
        # build a generator of column expressions to rename columns in a select
```

File: python/tempo/tsdf.py [hunk 3]
```diff
@@ -157,38 +155,39 @@ def __addPrefixToColumns(self, col_list: list[str], prefix: str) -> "TSDF":
...
+        rename_fn = (
+            lambda col: sfn.col(col).alias(col_map[col])
+            if col in col_map
+            else sfn.col(col)
         )
+        select_exprs = [rename_fn(col) for col in self.df.columns]
```
Comment (R7L208) [inline python/tempo/tsdf.py:172]:
```suggestion
        select_exprs = [
            sfn.col(col).alias(col_map[col] if col in col_map
            else sfn.col(col)
            for col in self.df.columns
        ]
```

Refactor to get rid of lambda and b/c this doesn't seem to be used a lot in other places in the code so probably not worth a function

**Review by R7L208 — CHANGES_REQUESTED** [review]
I suggested a few changes that make the code simpler. I'd like to move away from `lambda` unless it's actually needed

**Comment by R7L208** [comment]
Since column names are always strings or callable types, we may want to look at using Arrays instead of lists/tuples if we keep seeing issues with resolving operations on a large # of column names

File: python/tempo/tsdf.py (hunk 1 above)
Comment (tnixon) [reply python/tempo/tsdf.py:164]: I wanted to leave this in as a suggestion for the future. This is a nicer way to implement the rename, but only works in Spark 3.4+

Comment (tnixon) [reply python/tempo/tsdf.py:164]: I'll clarify in a comment

Comment (tnixon) [reply python/tempo/tsdf.py:164]:
```suggestion
        # In the future (when Spark 3.4+ is standard) we should implement batch rename using:
        # df = self.df.withColumnsRenamed(col_map)
```

File: python/tempo/tsdf.py [hunk 4]
```diff
@@ -161,15 +161,15 @@ def __addPrefixToColumns(self, col_list: list[str], prefix: str) -> "TSDF":
 
         # build a column rename map
         col_map = {col: "_".join([prefix, col]) for col in col_list}
+        # In the future (when Spark 3.4+ is standard) we should implement batch rename using:
```
Comment (R7L208) [inline python/tempo/tsdf.py:164]:
```suggestion
        # TODO - In the future (when Spark 3.4+ is standard) we should implement batch rename using:
```

Adding a TODO so it's easier to find again :)

**Review by R7L208 — APPROVED** [review]
NIT but suggested adding `TODO` on the comment regarding batch renaming so we can find it again easily.

Approving though. Thanks for the updates

**Pull request merged**